api/app/db/crud/weather_models.py

```diff
@@ -22,142 +21,132 @@
-        filter(WeatherStationModelPrediction.station_code.in_(station_codes)).\
-        filter(WeatherStationModelPrediction.prediction_timestamp >= start_date).\
-        filter(WeatherStationModelPrediction.prediction_timestamp <= end_date).\
-        filter(PredictionModelRunTimestamp.id ==
-               WeatherStationModelPrediction.prediction_model_run_timestamp_id).\
-        filter(PredictionModelRunTimestamp.prediction_model_id == PredictionModel.id,
-               PredictionModel.abbreviation == model).\
-        order_by(WeatherStationModelPrediction.station_code).\
-        order_by(WeatherStationModelPrediction.prediction_timestamp).\
-        order_by(PredictionModelRunTimestamp.prediction_run_timestamp.asc())
+    query = (
+        session.query(WeatherStationModelPrediction, PredictionModelRunTimestamp, PredictionModel)
+        .filter(WeatherStationModelPrediction.station_code.in_(station_codes))
+        .filter(WeatherStationModelPrediction.prediction_timestamp >= start_date)
+        .filter(WeatherStationModelPrediction.prediction_timestamp <= end_date)
+        .filter(PredictionModelRunTimestamp.id == WeatherStationModelPrediction.prediction_model_run_timestamp_id)
+        .filter(PredictionModelRunTimestamp.prediction_model_id == PredictionModel.id, PredictionModel.abbreviation == model)
+        .order_by(WeatherStationModelPrediction.station_code)
+        .order_by(WeatherStationModelPrediction.prediction_timestamp)
+        .order_by(PredictionModelRunTimestamp.prediction_run_timestamp.asc())
+    )
     return query
 
 
-def get_latest_station_model_prediction_per_day(session: Session,
-                                                station_codes: List[int],
-                                                model: str,
-                                                day_start: datetime.datetime,
-                                                day_end: datetime.datetime):
+def get_latest_model_predictions_subquery(session: Session, station_codes: List[int], day_start: datetime.datetime, day_end: datetime.datetime):
+    subquery = (
+        session.query(
+            func.max(WeatherStationModelPrediction.prediction_timestamp).label("latest_prediction"),
+            WeatherStationModelPrediction.station_code,
+            func.date(WeatherStationModelPrediction.prediction_timestamp).label("unique_day"),
+        )
+        .filter(
+            WeatherStationModelPrediction.station_code.in_(station_codes),
+            WeatherStationModelPrediction.prediction_timestamp >= day_start,
+            WeatherStationModelPrediction.prediction_timestamp <= day_end,
+            func.date_part("hour", WeatherStationModelPrediction.prediction_timestamp) == 20,
+        )
+        .group_by(WeatherStationModelPrediction.station_code, func.date(WeatherStationModelPrediction.prediction_timestamp).label("unique_day"))
```

Comment:
Is this `group_by` clause necessary? Just wondering if groups in a subquery are reflected in the final query that uses the subquery.

Reply:
Oh, I see it was part of the original subquery. nvm

Reply:
It's required because `get_latest_model_predictions_subquery` uses an aggregate function (`func.max(WeatherStationModelPrediction.prediction_timestamp).label("latest_prediction")`)